opflexagent/firewall_wrapper.py

```diff
@@ -0,0 +1,115 @@
+
+    This wrapper class keeps a map of drivers to use for
+    implementing a firewall driver. The firewall driver methods
+    are divided into two groups:
+      1) methods where only one driver implementation should be called
+      2) methods where all driver implementations should be called
+
+    The first group is called when there is only a single argument:
+    the neutron port. The second group is called in all other cases.
+    The first group uses a key to determine which driver gets invoked.
+    The key is obtained by a user-defined strategy function, which gets
+    passed the port object. The strategy function is meant to perform
+    some operation (e.g. search) on the port to determine and return
+    the key to use for looking up the appropriate firewall driver.
+
+    There is also a defaulit firewall driver. This is used when there
+    is only a single firewall driver is used, and therefore a key is
+    not needed.
+    """
+
+    def __init__(self):
+        self._strategy_map = {}
+        self._strategy_fn = None
+        self._default_strategy = None
+        self._port_methods = ['prepare_port_filter',
+                              'apply_port_filter',
+                              'update_port_filter',
+                              'remove_port_filter',
+                              'remove_port_filter']
+        self._all_strategy_methods = ['filter_defer_apply_on',
+                                      'filter_defer_apply_off',
+                                      'ports',
+                                      'defer_apply',
+                                      'update_security_group_members',
+                                      'update_security_group_rules']
+
+    def _do_port_strategy(self, name, port):
+        """Call a single firewall driver method"""
+        strategy_obj = self._default_strategy
+        if self._strategy_map and self._strategy_fn:
+            key = self._strategy_fn(port, self._strategy_map)
+            if key:
+                strategy_obj = self._strategy_map[key]
+        return getattr(strategy_obj, name)(port)
+
+    def _do_all_strategy(self, name, *s_args, **s_kwargs):
+        """Call all the firewall driver methods"""
+        if self._default_strategy:
+            return getattr(self._default_strategy, name)(*s_args, **s_kwargs)
+        else:
+            ret_list = []
+            for driver in self._strategy_map.values():
+                ret_list.append(getattr(driver, name)(*s_args, **s_kwargs))
+            return ret_list
+
+    def __getattr__(self, name):
+        def _do_strategy(*args, **kwargs):
+            """Invoke the appropiate strategy method
+
+            This checks the number of arguments, and if 1,
+            invokes the key-based strategy. If not 1, it calls
+            all the drivers.
+            """
+            if name in self._port_methods:
+                return self._do_port_strategy(name, args[0])
```

Comment:
you don't know if there's only one argument in this case, should still pass *args and **kwargs